src/compiler/checker.ts

```diff
@@ -10212,7 +10217,7 @@ namespace ts {
             }
 
             // Functions with with an explicitly specified 'void' or 'any' return type don't need any return expressions.
-            if (returnType === voidType || isTypeAny(returnType)) {
+            if (returnType === voidType || isTypeAny(returnType) || isUnionContaining(returnType, TypeFlags.Any) || isUnionContaining(returnType, TypeFlags.Void)) {

```

Comment:
I think we can rewrite this to
```ts
[old code] || (returnType.flags & TypeFlags.Union && someConstituentTypeHasKind(type, TypeFlags.Any | TypeFlags.Void)
```
to avoid having a new helper function